module/zfs/metaslab.c

```diff
@@ -272,6 +272,12 @@ uint64_t metaslab_trace_max_entries = 5000;
  */
 int max_disabled_ms = 3;
 
+/*
+ * Time (in seconds) to respect ms_max_size when the metaslab is not loaded.
+ * To avoid 64-bit overflow, don't set above UINT32_MAX.
+ */
+unsigned long max_size_cache_sec = 3600; /* 1 hour */
```

Comment:
For consistency let's prefix this tunable with `zfs_metaslab_` like (most of) the other metaslab related tunings.

>  A more sophisticated algorithm here might be helpful, but this works reasonably well.

Based on your comment in the PR description it sounds like this value is somewhat arbitrary.  As I understand it, the important thing is that eventually the cached value is considered stale after a certain amount of activity (frees).  My concern would be that a wall clock time isn't a great approximation for this.  A slightly better heuristic could be to specify this value in terms of txgs processed since the metaslab was unloaded.  If 3600 seconds was determined a good value then it could instead be specified as 720 txgs (assuming the default 5 seconds per-txg target value).  Or is there some other reason why it's useful to specify this in seconds?

Reply:
> My concern would be that a wall clock time isn't a great approximation for this. A slightly better heuristic could be to specify this value in terms of txgs processed since the metaslab was unloaded. 

How would specifying it in TXG's be better?  I find wall-clock time easier to reason about, because TXG's don't flow at an even rate on all systems.

Reply:
> TXG's don't flow at an even rate on all systems.

This is my point.  Since TXGs don't flow at an even rate specifying a wall clock time tells you little about how much churn the pool has experienced, and consequently how many frees have accumulated.  TXGs certainly aren't ideal either, but they do at least track the overall pool activity.  I suppose ideally we'd count the frees and use that.

That said, I don't feel too strongly about this.  And since this version has seen the most testing I'm content to leave it as seconds.  We can always revisit and replace it with a better algorithm if needed.

Reply:
We could definitely be smarter about what we're trying to do here, which is to limit the inaccuracy of `ms_max_size`.  IIRC @pcd1193182 tried some more sophisticated algorithms, which were often better but were also much harder to reason about and be sure that they never did something egregiously wrong.

Reply:
I implemented an algorithm that was based on a statistical model of the free chunk sizes; as frees were issued to a metaslab the estimated max_size was slowly increased.  Once the estimated max_size was more than a certain amount larger than the cached max_size, we started using it instead.  My initial analysis found that we could pretty reliably get the actual max size to be at least as large as the expected max size 95% of the time while still providing meaningful aging rates. However, as Matt said, this algorithm's dynamic behavior is not trivial to reason about.

Reply:
I see, OK.  I'm convinced!